cosmos/operators/watcher.py

```diff
@@ -90,3 +118,132 @@
+      get per-model updates until the build completes.
+
+    This keeps the heavy dbt work centralised while providing near real-time
+    feedback and granular task-level observability downstream.
+    """
+
+    def __init__(
+        self,
+        *,
+        render_config: RenderConfig | None = None,
+        **kwargs: Any,
+    ) -> None:
+        # Store so we can honour select/exclude when building flags
+        self.render_config: RenderConfig | None = render_config
+
+        task_id = kwargs.pop("task_id", "dbt_build_coordinator")
+        super().__init__(task_id=task_id, **kwargs)
+
+    def add_cmd_flags(self) -> list[str]:
+        flags: list[str] = super().add_cmd_flags()
+
+        self.log.info("DbtBuildCoordinatorOperator: render_config: %s", self.render_config)
+        if self.render_config is not None and self.render_config.exclude:
+            flags.extend(["--exclude", *self.render_config.exclude])
+        if self.render_config is not None and self.render_config.select:
+            flags.extend(["--select", *self.render_config.select])
+        return flags
+
+    @staticmethod
+    def _serialize_event(ev: EventMsg) -> dict[str, Any]:
+        """Convert structured dbt EventMsg to plain dict."""
+        from google.protobuf.json_format import MessageToDict
+
+        return MessageToDict(ev, preserving_proto_field_name=True)  # type: ignore[no-any-return]
+
+    def _handle_startup_event(self, ev: EventMsg, startup_events: list[dict[str, Any]]) -> None:
+        info = ev.info  # type: ignore[attr-defined]
+        raw_ts = getattr(info, "ts", None)
+        ts_val = raw_ts.ToJsonString() if hasattr(raw_ts, "ToJsonString") else str(raw_ts)  # type: ignore[union-attr]
+        startup_events.append({"name": info.name, "msg": info.msg, "ts": ts_val})
+
+    def _handle_node_finished(
+        self,
+        ev: EventMsg,
+        context: Context,
+    ) -> None:
+        self.log.debug("DbtBuildCoordinatorOperator: handling node finished event: %s", ev)
+        ti = context["ti"]
+        uid = ev.data.node_info.unique_id
+        ev_dict = self._serialize_event(ev)
+        payload = base64.b64encode(gzip.compress(json.dumps(ev_dict).encode())).decode()
+        ti.xcom_push(key=f"nodefinished_{uid.replace('.', '__')}", value=payload)
+
+    @contextmanager
+    def _patch_runner(self, callback: Callable[[EventMsg], None]) -> Any:
+        import cosmos.dbt.runner as _dbt_runner_mod
+
+        original = _dbt_runner_mod.get_runner
+
+        def _patched_get_runner() -> Any:
+            from dbt.cli.main import dbtRunner
+
+            return dbtRunner(callbacks=[callback])
+
+        _dbt_runner_mod.get_runner = _patched_get_runner  # type: ignore[assignment]
+        if hasattr(original, "cache_clear"):
+            # Clear the cache of the original get_runner function to ensure that
+            # no stale cached runners are used after monkey-patching. This prevents
```

Comment:
Do we really need monkey-patching in this execution mode when using `dbtRunner`?

Reply:
yes, we're registering the callback function by patching the dbtRunner.

Reply:
yes removed this monkeypatching and added capability to base get_runner to accept callbacks